### ISSUE TITLE
Add `largest` and `smallest` to the `WhereOperatorSelectionMode` Enum 

### ISSUE THREAD

**Issue by iRon7** [PowerShell/PowerShell#20599]
### Summary of the new feature / enhancement

There are currently several ways to [Query PSCustomObject Array for row with largest value](https://stackoverflow.com/questions/64344832/query-pscustomobject-array-for-row-with-largest-value).

But they are either pretty verbose or expensive

```PowerShell
$array = @(
    [PSCustomObject]@{Name='test1';Value=105}
    [PSCustomObject]@{Name='test2';Value=101}
    [PSCustomObject]@{Name='test3';Value=512}
)
```

### Proposed technical implementation details (optional)

Adding a `largest` - and `smallest` to the [`WhereOperatorSelectionMode` Enum](https://learn.microsoft.com/dotnet/api/system.management.automation.whereoperatorselectionmode) might be more effective:

Wishful thinking:
```PowerShell
$array.where({ $_.Value }, 'Largest')
```

**Footnote**
* Unlike the default PowerShell behavior for string comparison, I think it (if accepted at all) should be case sensitive as an case insensitive comparison could be reached using the `.ToLower()` - or  `.ToUpper()` method, e.g.:

```PowerShell
$array.where({ $_.Name.ToLower() }, 'Largest')
```

**Comment by mklement0** [comment]
Honestly, I don't think this fits in well with how the `.Where()` method currently works: It does _per-object_ filtering, and the script block's output is invariably treated as a _Boolean_.

Given that `Sort-Object` now has `-Top` and `-Bottom` parameters, the following seems succinct enough for me:

```powershell
$array = @(
    [PSCustomObject]@{Name='test1';Value=105}
    [PSCustomObject]@{Name='test2';Value=101}
    [PSCustomObject]@{Name='test3';Value=512}
)

# Get minimum
$array | Sort-Object -Top 1 -Property Value
# Get maximum
$array | Sort-Object -Bottom 1 -Property Value
```

That said, what `Sort-Object` currently _doesn't_ support is:

* Getting _all_ minima and maxima (should there be multiple). (Conceivably, `-AllTop` and `-AllBottom` switches could be introduced).

* Allowing you to find _both_ the minimum and maximum in a single operation, because `-Top` and `-Bottom` are in different parameter sets. (By contrast, `Measure-Object` allows specifying both `-Minimum` and `-Maximum`; granted, retrieving _multiple_ top and bottom values can make it harder to separate the maxima from the minima; this could either be left to the user or solved with an ETS property).

In other words: Perhaps it's worth creating `Sort-Object` feature requests to overcome these limitations.

**Comment by iRon7** [comment]
@mklement0,

Thank you for the response.

> Given that Sort-Object now has `-Top` and `-Bottom` parameters, the following seems succinct enough for me:

I need to admit that I was not aware of these `-Top` and `-Bottom` parameters which makes this proposal less interesting from a syntactical point of view.
Yet, I still think that sorting a list of objects and than select the top/bottom (or minimum/maximum) is presumably not effective (in memory usage and performance) as for the sorting part you will need to hold each item against multiple other items (for any of the existing [sorting algorithm](https://en.wikipedia.org/wiki/Sorting_algorithm)) where for a `where` method you only need to hold each item against one item (the -currently- largest/smallest item) and update it if required.

Not sure what is a fair comparison using `Foreach-Object` the optimized `Foreach-ObjectFast` (#10982). Anyways the `foreach` statement is about twice as fast for last lists and I presume a comparable result for a `where` largest/smallest method.

```PowerShell
$Size = 100000
$List = 1..$Size | Foreach-Object { [pscustomobject]@{ Name = "Name$_"; Value = [Math]::Abs($_ / 2 - $Size) } }

@{ 'Sort-Object' = Measure-command {
    $List | Sort-Object -Top 1 -Property Value
}}

@{ 'Foreach' = Measure-command {
    $Top = $Null
    foreach ($Item in $List) {
        if ($Null -eq $Top -or $Item.Value -lt $Top.Value) { $Top = $Item }
    }
    $Top
}}

@{ 'ForEach-Object' = Measure-command {
    $List | ForEach-Object { $Top = $Null } {
        if ($Null -eq $Top -or $_.Value -lt $Top.Value) { $Top = $_ }
    } { $Top }
}}

@{ 'ForEach-ObjectFast' = Measure-command {
    $List | ForEach-ObjectFast -Begin { $Top = $Null } -Process {
        if ($Null -eq $Top -or $_.Value -lt $Top.Value) { $Top = $_ }
    } -end { $Top }
}}
```

```PowerShell
Name                           Value
----                           -----
Sort-Object                    00:00:00.9188725
Foreach                        00:00:00.4545986
ForEach-Object                 00:00:01.0624101
ForEach-ObjectFast             00:00:00.5842748
```

**Comment by mklement0** [comment]
Good point about `Sort-Object` of necessity collecting all input.
(`-Top/-Bottom 1` and possibly `-AllTop` / `-AllBottom` at least technically could be _special-cased_, but that would increase implementation complexity quite a bit).

I do think that having an efficient way to find minima and maxima would be helpful, but, as noted, I think `.Where()` is an awkward place to put it - but perhaps others disagree; at the very least, for symmetry with `Measure-Object`, the enumeration values should be `Minimum` and `Maximum`

Note that a `.Where()` solution - while generally better-performing than a _cmdlet_ solution - precludes _streaming_, both on input and output: the input collection must be collected in full before `.Where()` can be called, and the output is a collection instance.

Another possibility that _would_ support streaming (not necessarily mutually exclusive with  your suggestion, though it wouldn't make for an obvious correspondence between the expression and the cmdlet option) would be to enhance `Measure-Object` with a `-PassThru` option that passes the minimum / maximum _input object through_; what makes this slightly awkward is that `-PassThru` _only_ makes sense with `-Minimum` and `-Maximum`, not any of `Measure-Object`'s other switches.

**Comment by iRon7** [comment]
@mklement0, again, thanks for the comments.

>for symmetry with Measure-Object, the enumeration values should be `Minimum` and `Maximum`

Agree, I have changed this in title and the initial proposal.

> I think `.Where()` is an awkward place to put it

Maybe the enumeration values should be even `isMinimum` and `isMaximum` which betters shows some linguistic elegance of the statement `$array.where({ $_.Value }, 'isMaximum'` → **where-array-value-is-maximum**.
